doc/cbf/cbf_hld.md

```diff
@@ -0,0 +1,215 @@
+
+In order to support this mapping, 2 new mapping tables will be added to the CONFIG_DB for the DSCP/EXP to FC mapping and a new CLASS_BASED_NEXT_HOP_GROUP table will be added to APPL_DB to support the new FC-aware next hop groups.
+
+# 2 Requirement Overview
+## 2.1 Functional Requirements
+
+Allow traffic to be forwarded through the network based on their DSCP/EXP values following these rules:
+- If a packet is not matched against an FC value and the route for its destination does not reference a CBF NHG, the packet will use the route's NH
+- If a packet is not matched against an FC value and the route for its destination references a CBF NHG, the packet will be dropped
+- If a packet is matched against an FC value and the route for its destination does not reference a CBF NHG, the packet will use the route's NH
+- If a packet is matched against an FC value and the route for its destination references a CBF NHG which maps the packet's FC value, the packet will use the mapped NHG
+- If a packet is matched against an FC value and the route for its destination references a CBF NHG which doesn't map the packet's FC value, the packet will be dropped
+
+## 2.2 Configuration and Management Requirements
+- DSCP/EXP to FC maps must be allowed to be configured via the 2 CONFIG_DB tables with no requirement to be configurable via CLI.
+
+## 2.3 Scalability Requirements
+- Unchanged.
+
+## 2.4 Warm Boot Requirements
+- Unchanged - the new class based next hop group table must be compatible with existing warm boot requirements.
+
+## 2.5 Restrictions
+- fpmsyncd is not updated to use the new CLASS_BASED_NEXT_HOP_GROUP_TABLE as part of this enhancement. Anyone wishing to use this feature must use a modified version of fpmsyncd, or program the table directly.
+
+# 3 Design
+## 3.1 Overview
+This design directly changes CONFIG_DB, APPL_DB, orchagent and sairedis.
+
+## 3.2 DB Changes
+### 3.2.1 APPL DB
+Based on the next hop group split (https://github.com/Azure/SONiC/pull/712) on which this HLD is based on, a new CLASS_BASED_NEXT_HOP_GROUP table will be added to the APPL_DB with the following format:
+```
+### CLASS_BASED_NEXT_HOP_GROUP_TABLE
+    ;Stores a list of FC-aware next hop groups.
+    ;Status: Mandatory
+    key           = CLASS_BASED_NEXT_HOP_GROUP_TABLE:string ; arbitrary string identifying the class based next hop group, as determined by the programming application.
+    members       = NEXT_HOP_GROUP_TABLE.key,    ; one or more indexes within NEXT_HOP_GROUP_TABLE, separated by “,”
+    class_map     = number:number,   ; one or more mapping from Forwarding Class to index in "members" field to use as NHG, separated by ","
+```
+
+Example:
+    127.0.0.1:6379[1]> hgetall  "CLASS_BASED_NEXT_HOP_GROUP:CbfNhg1"
+    1) "members"
+    2) "Nhg1,Nhg2,Nhg3,Nhg4"
+    3) "class_map"
+    4) "0:0,1:0,2:1,3:1,4:2,5:2,6:3,7:3"
```

Comment:
Is it possible to define one more table for class (or DSCP/EXP) to NH mapping and refer it from this table?

Reply:
> Is it possible to define one more table for class (or DSCP/EXP) to NH mapping and refer it from this table to make the mapping clean?

I too second that. As discussed in the community meeting consider having a separate table for DSCP_to_FC, and refer to those table names in CLASS_BASED_NEXT_HOP_GROUP_TABLE.


```
### FC_TO_NH_INDEX_MAP_TABLE
; FC to Next hop group index map
key                    = "FC_TO_NH_INDEX_MAP_TABLE:"name
;field
fc_num = 1*DIGIT
;value
nh_index  = 1*DIGIT;  index of NH inside NH group

### CLASS_BASED_NEXT_HOP_GROUP_TABLE
    ;Stores a list of FC-aware next hop groups.
    ;Status: Mandatory
    key           = CLASS_BASED_NEXT_HOP_GROUP_TABLE:string ; arbitrary string identifying the class based next hop group, as determined by the programming application.
    members       = NEXT_HOP_GROUP_TABLE.key,    ; one or more indexes within NEXT_HOP_GROUP_TABLE, separated by “,”
    map_forwarding_class_to_index    = ref_hash_key_reference   ; refer
```

Reply:
Could you explain the reasoning behind preferring the separate map? I see that it matches the style of other similar maps, but would like to understand why those were done that way originally.